petastorm/pytorch.py

```diff
@@ -423,3 +367,139 @@ def __enter__(self):
+            buffer_full = True
+            last_row = rows_capacity-n_rows
+            expected_rows = rows_capacity
+        if buffer is None:
+            # Initialize buffer as a list of empty tensors
+            buffer = []
+            for v in items:
+                buffer.append(torch.empty((rows_capacity,) + v.shape[1:], dtype=v.dtype, device=v.device))
+        # Copy new items into buffer
+        for i, v in enumerate(items):
+            buffer[i][n_rows:expected_rows] = v[:last_row]
+        n_rows = expected_rows
+
+    # At this point, dataloader has enough rows storted in memory.
+    # Stop the reader rather than draining the remainder of reader.
+    # If reader has infinite epochs, draining will be deadlock.
+    reader.stop()
+    # Truncate empty tensors if capacity is larger than total rows.
+    if n_rows < rows_capacity:
+        for i, v in enumerate(buffer):
+            buffer[i] = buffer[i][:n_rows]
+    return (keys, buffer)
+
+
+class InMemBatchedDataLoader(object):
+    """
+    Same as BatchedDataLoader except it only loads upto capacity rows into memory.
+    This class doesn't allow to be used multiple-times, so it doesn't inherit LoaderBase.
+    """
+
+    def __init__(self, reader, batch_size=1,
+                 transform_fn=None,
+                 num_epochs=1,
+                 seed=0,
+                 rows_capacity=1024,
+                 shuffle=False):
+        """
+        Initializes a data loader object.
+        This class does not support special types that are not supported in PyTorch (decimal/string).
+        :param reader: petastorm Reader instance
+        :param batch_size: the number of items to return per batch; factored into the len() of this reader
+        :param transform_fn: an optional callable to convert batches from the reader to PyTorch tensors
+        :param num_epochs: number of epochs.
+        :param seed: random seed used to shuffle.
+        :param rows_capacity: number of rows to be loaded into memory.
+        :param shuffle: If ``True``, indices will be shuffled in every epoch.
+        """
+        super(InMemBatchedDataLoader, self).__init__()
+        self._reader = reader
+        self._batch_size = batch_size
+        self._transform_fn = transform_fn or torch.as_tensor
+        self._num_epochs = num_epochs
+        self._seed = seed
+        self._rows_capacity = rows_capacity
+        self._shuffle = shuffle
+        self._in_iter = False
+
+    def __iter__(self):
+        """
+        The Data Loader iterator stops the for-loop when num_epochs is reached.
+        """
+        if self._in_iter:
+            raise RuntimeError("InMemBatchedDataLoader couldn't be used multiple times, please\
+                    specify total number of epochs using num_epochs in constructor.")
+        self._in_iter = True
+        try:
+            keys, buffer = _load_rows_into_mem(self._reader, self._transform_fn, self._rows_capacity)
```

Comment:
Consider adding a bit more comments describing structure of `keys` and `buffer`. I think it may help a reader understand the code better before digging into `_load_rows_into_mem` implementation.